### PR TITLE
Hide 'Mirror Settings' when unneeded, improve hints

### PR DESCRIPTION
**Pull request by n0toose**

### CODE REVIEW

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]: These will invalidate all current translations. I think we better move at minimum the URL out of the translations, ideally the whole `<a>` tag.

**Comment by n0toose** [comment]
This change seems small, but it was actually quite exhausting figuring out the logic, as well as figuring out how to communicate those details to the user properly. It implements a change suggested in a Codeberg ticket: https://codeberg.org/Codeberg/Community/issues/1005

I believe that this is ready for review, but I still need to go over every single case, just to be 100% sure. I am mildly concerned about the scalability of this change.

Authoring this change took me ~90 minutes.

File: options/locale/locale_en-US.ini
Comment (silverwind) [reply options/locale/locale_en-US.ini]: Some ideas for this in https://github.com/go-gitea/gitea/pull/24426#discussion_r1181111566, but nothing concrete yet on how we should do this.

Comment (n0toose) [reply options/locale/locale_en-US.ini]: > These will invalidate all current translations.

I get that it's effort and that extra effort from the translators would be required, but intentional. I'm assuming that the TM would help here.

I am not sure if you are saying that "invalidating all current translations is unacceptable" or that we should should take advantage of the opportunity to remove the `<a>` tag from the translations.

> I think we better move at minimum the URL out of the translations, ideally the whole <a> tag.

My change proposes that we redirect the user directly to the right section. `en-us` is another problem, which is why I didn't do that in the first place... are there other examples of that happening?

File: templates/repo/settings/options.tmpl
Comment (n0toose) [inline templates/repo/settings/options.tmpl]: Potential logic error here: If mirrors are disabled, the Mirror Settings should not be shown to the user, even if their repository *was* a mirror, because it *doesn't serve the function of a mirror anymore*, if I understood this correctly.

It could be the case that mirrors are enabled, but no new push and pull mirrors are allowed and the repository is *not* a mirror. ~~I think that the first `or` should be replaced with an `and`.~~ EDIT: Nope, mirror settings should still be modifiable, even if new mirrors are disabled.

Comment (n0toose) [reply templates/repo/settings/options.tmpl]: It seems like a very trivial problem to fix, but I'll take a break for now.

Comment (n0toose) [reply templates/repo/settings/options.tmpl]: Wrong variable was used.

**Comment by n0toose** [comment]
I got super confused at first because I falsely assumed that `DisableNewPullMirrors` was added to `ctx.Data` and ended up doing things such as using the wrong variables.

image

image

image

For some reason, I can't reach the last condition by setting, even after having set a mirror...

```
[mirror]
; ENABLED            = true;
DISABLE_NEW_PUSH   = true;
DISABLE_NEW_PULL   = true;
```

... because the menu disappears even if the repository should (???) be a mirror.

I'm also concerned about performance and i18n.

File: options/locale/locale_en-US.ini
Comment (silverwind) [reply options/locale/locale_en-US.ini]: Thanks, I see you moved out the URLs, which is great.

My point is, take a look at percentages on https://crowdin.com/project/gitea. Some of these languages will probably take a very long time to receive an update, if at all. So if they had translated the old links in the past, they could stay on them for years to come in theory.

Comment (n0toose) [reply options/locale/locale_en-US.ini]: Got it. Are you sure that it's going to be i18n-able?

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
settings.mirror_settings.docs.no_new_mirrors = Your repository is mirroring changes to or from another repository. Please keep in mind that you can't create any new mirrors at this time.
```

Assuming this `or` is correct here, I like it better than slash.

**Comment by silverwind** [comment]
> For some reason, I can't reach the last condition by setting, even after having set a mirror...

I guess just fake it in the template condition to validate the rendering.

Are those screenshots above up to date? That bold text in the second one looks odd if it's still there, I'd remove it.

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
settings.mirror_settings.docs = Set up your repository to automatically synchronize commits, tags and branches with another repository.
```

Bit shorter and says the same basically.

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
settings.mirror_settings.docs.disabled_pull_mirror.instructions = Set up your project to automatically push commits, tags and branches to another repository. Pull mirrors have been disabled by your site administrator.
```

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
settings.mirror_settings.docs.disabled_push_mirror.instructions = Set up your project to automatically pull commits, tags and branches from another repository.
```

**Comment by n0toose** [comment]
> Are those screenshots above up to date? That bold text in the second one looks odd if it's still there, I'd remove it.

Will address this later.

**Review by silverwind — APPROVED** [review]

File: templates/repo/settings/options.tmpl
Comment (lunny) [inline templates/repo/settings/options.tmpl]: Why it's `or` but not `and`?

Comment (n0toose) [reply templates/repo/settings/options.tmpl]: I'll write this in free text:

If the repository is a mirror OR new pull mirrors and new push mirrors aren't allowed, then the repository's settings will be shown to the user.

Otherwise, changing the mirror settings for repositories that were created *before* hypothetically disabling the mirrors would not be allowed. `.MirrorsEnabled` effectively disables new pull mirrors and new push mirrors, so I thought that this condition would be reasonable.

If I am not wrong (there's a chance that I am, I don't remember if I managed to verify this or not), disabling mirrors removes the mirror status. This needs to be checked again.

Comment (n0toose) [reply templates/repo/settings/options.tmpl]: > If I am not wrong (there's a chance that I am, I don't remember if I managed to verify this or not), disabling mirrors removes the mirror status. This needs to be checked again.

What I said earlier seems to be wrong (Sorry, I "bruteforced" this change until it worked and forgot many of the details right afterwards).

Should the settings remain modifiable from the web UI if the repository is marked as a mirror and mirrors are disabled?

Comment (lunny) [reply templates/repo/settings/options.tmpl]: If it's a mirror repository, it will be displayed. Even if the mirror is disabled after the mirror repository is created. Although the mirroring action will be stopped because the mirroring is disabled. So maybe it could still be editable but maybe we should have a remind that mirroring is disabled, this mirror repository will not be updated.

Comment (n0toose) [reply templates/repo/settings/options.tmpl]: Excellent idea, will implement later.

Comment (n0toose) [reply templates/repo/settings/options.tmpl]: Disabling .MirrorsEnabled results in the change - implemented as you described it - not being visible at all. I can't spot any relevant code excerpts, however, I believe that my condition is otherwise correct.

image

**Comment by wxiaoguang** [comment]
Maybe there could be some variables to help to simplify the logic:

```
{{$enableNewPullMirros := not .DisableNewPullMirrors}}
{{$enableNewPushMirros := not .DisableNewPushMirrors}}

{{if or $enableNewPullMirros $enableNewPushMirros .Repository.IsMirror}}
  ...
{{end}}

{{if and $enableNewPullMirros $enableNewPushMirros}}
  ...
{{else if $enableNewPullMirros}}
  ...
{{else if $enableNewPushMirros}}
  ...
{{else}}
  ...
{{end}}
```

**Comment by n0toose** [comment]
> Maybe there could be some variables to help to simplify the logic

Just to be clear, as I am juggling around a bunch of Gitea PRs right now and some of them are more problematic than others - for the sake of setting my priorities straight: Is this an instruction or a recommendation?

**Comment by wxiaoguang** [comment]
> Is this an instruction (as in, "do this and we'll evaluate whether we can merge this after you do so") or a recommendation?

Not "we", but only "me", I don't represent the community. Other maintainers could have other options. It is more a personal option / recommendation from me. Even if I approve it doesn't mean there is no block from others ......

According to Gitea's review habits (I think there is a new contributing document now?), only serious problems get "requested changes / block" mark.

My personal option about the reviews (only to share, only my option):

1. I only approve PRs I can >90% sure and understand. If I approved PR has bugs and the author doesn't have time to follow, I will also fix, I consider such problem is also my bug.
2. Review comments could always be discussed and argued. My PRs also receive a lot of review comments, some useful reviews I takes, some unrelated/misunderstood/non-feasible reviews I would explain with real cases/examples, or why I don't like it.

**Comment by n0toose** [comment]
> I think it can use some clearer logic, it also helps other reviewers and future developers. That's just only my opinion, if others like the current approach, I do not have objection.

Got it, thank you for the clarification. I didn't think of my code being very complicated - I assumed that the thing that made my change complicated was that it sometimes doesn't behave the way one would expect it to behave because variable behavior is not explicitly defined by the config itself. I'll take a look at improving that.

**Comment by n0toose** [comment]
- I did not test the recent set of changes extensively, under the assumption that no logic-related behavior should change. There was one exception (I got an error when everything was enabled) because I mixed up an `and` with an `or`. Checked once again that the conditions were right.
- I preserved the `(not .DisableNewPullMirrors)` stuff and decided to reduce the level of complexity elsewhere. I think it improves maintainability.

This definitely feels better for me. Hope the result looks better to you as well, thanks for the directions.

**Review by wxiaoguang — APPROVED** [review]
Overwall LGTM

The doc site has been migrated to gitea.com. Maybe some maintainers with writer permission could help to improve.

**Comment by n0toose** [comment]
Doing some final checks...

**Comment by n0toose** [comment]
Minor errors spotted, please do not merge yet.

**Comment by n0toose** [comment]
Mirrors enabled, no configuration set for disabling new push and pull mirrors:

image

Mirrors enabled, pull mirrors disabled:

image

Mirrors enabled, push mirrors disabled (with existing push mirror):

image

Mirrors enabled, new push/pull mirrors enabled/disabled (with errors) - this also shows when mirrors entirely disabled:

image

**Comment by n0toose** [comment]
Some strings were broken and fixed. I previously wrote some strings because I made some wrong assumptions about the behavior, and I also had to make a few updates after figuring out that a repository had the `.Repository.IsMirror` boolean set to true only when it actively managed to mirror things.

I made a change so that the menu would appear when mirrors are entirely disabled and configured push/pull mirrors exist, even if they do not work. **I am very skeptical of that change,** and I realize that these changes came after the approval of the maintainers and *effectively circumvent the review process.* Really sorry about this, I just kept making false assumptions about the behavior of this feature as there were a few seemingly unexplainable inconsistencies. I added some comments so that people in the future won't fall for the same fallacies that I did.

A last, final check **before merging** and "OK" would be great.

**Comment by n0toose** [comment]
> The doc site has been migrated to gitea.com. Maybe some maintainers with writer permission could help to improve.

I did not manage to address this.

**Pull request merged**
